chainindex/read.go

```diff
@@ -0,0 +1,148 @@
+package chainindex
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+const headIndexedWaitTimeout = 5 * time.Second
+
+func (si *SqliteIndexer) GetCidFromHash(ctx context.Context, txHash ethtypes.EthHash) (cid.Cid, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return cid.Undef, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var msgCidBytes []byte
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
+		return si.queryMsgCidFromEthHash(ctx, txHash, &msgCidBytes)
+	}); err != nil {
+		return cid.Undef, err
+	}
+
+	msgCid, err := cid.Cast(msgCidBytes)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to cast message CID: %w", err)
+	}
+
+	return msgCid, nil
+}
+
+func (si *SqliteIndexer) queryMsgCidFromEthHash(ctx context.Context, txHash ethtypes.EthHash, msgCidBytes *[]byte) error {
+	return si.getMsgCidFromEthHashStmt.QueryRowContext(ctx, txHash.String()).Scan(msgCidBytes)
+}
+
+func (si *SqliteIndexer) GetMsgInfo(ctx context.Context, messageCid cid.Cid) (*MsgInfo, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return nil, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var tipsetKeyCidBytes []byte
+	var height int64
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
+		return si.queryMsgInfo(ctx, messageCid, &tipsetKeyCidBytes, &height)
+	}); err != nil {
+		return nil, err
+	}
+
+	tipsetKey, err := cid.Cast(tipsetKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	return &MsgInfo{
+		Message: messageCid,
+		TipSet:  tipsetKey,
+		Epoch:   abi.ChainEpoch(height),
+	}, nil
+}
+
+// This function attempts to read data using the provided readFunc.
+// If the initial read returns no rows, it waits for the head to be indexed
+// and tries again. This ensures that the most up-to-date data is checked.
+// If no data is found after the second attempt, it returns ErrNotFound.
+func (si *SqliteIndexer) readWithHeadIndexWait(ctx context.Context, readFunc func() error) error {
+	err := readFunc()
+	if err == sql.ErrNoRows {
+		// not found, but may be in latest head, so wait for it and check again
+		if err := si.waitTillHeadIndexed(ctx); err != nil {
+			return xerrors.Errorf("failed while waiting for head to be indexed: %w", err)
+		}
+		err = readFunc()
+	}
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return ErrNotFound
+		}
+		return xerrors.Errorf("failed to get message info: %w", err)
```

Comment:
msg not generic enough - you're also using this for cid hash

Reply:
Fixed.